### PR TITLE
crimson/osd/pg: implement PG::PGLogEntryHandler::remove()

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/67564
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/pg.cc [hunk 1]
```diff
@@ -1771,4 +1771,21 @@ PG::already_complete(const osd_reqid_t& reqid)
...
+    auto &snap_mapper = pg->get_snap_mapper();
+    int r = snap_mapper.remove_oid(
+      soid,
+      &_t);
+    if (!(r == 0 || r == -ENOENT)) {
+      logger().debug("{}: remove_oid returned {}", __func__, cpp_strerror(r));
+      ceph_abort();
+    }
+  }
+}
```
Comment (Matan-B) [inline src/crimson/osd/pg.cc:1789]: Can we please move this into `PG::clear_object_snap_mapping`?

File: src/crimson/osd/pg_backend.cc [hunk 2]
```diff
@@ -995,6 +995,12 @@ PGBackend::create_iertr::future<> PGBackend::create(
   return seastar::now();
 }
 
+void PGBackend::remove(const hobject_t &hoid, ceph::os::Transaction &txn)
+{
+  logger().debug("{}: remove {} on pglog rollback", __func__, hoid);
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/osd/pg_backend.cc:1000]: Yeah, will do.

File: src/crimson/osd/pg.cc (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc:1789]: I think we need `remove_snap_mapped_object`. `clear_object_snap_mapping` is always used with `ceph::os::Transaction::remove`, which can be replace by `remove_snap_mapped_object`. What do you think?

Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc]: All coments addressed, please take a look again:-)

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
jenkins test api

File: src/crimson/osd/pg.cc [hunk 3]
```diff
@@ -1771,4 +1771,20 @@ PG::already_complete(const osd_reqid_t& reqid)
...
 }
 
+void PG::remove_object_maybe_snapmapped(
+  ceph::os::Transaction &t,
+  const hobject_t &soid)
+{
+  t.remove(
+    coll_ref->get_cid(),
+    ghobject_t{soid, ghobject_t::NO_GEN, pg_whoami.shard});
+  if (soid.snap < CEPH_MAXSNAP) {
```
Comment (Matan-B) [inline src/crimson/osd/pg.cc:1784]: I'm ok with uniting `remove_snap_mapped_object` and `clear_object_snap_mapping`. Naming suggestion below, up to you :)
```
PG::remove_snap_mapped_object(
  ..
  ..
  // clear object snap mapping
  if (soid.snap < CEPH_MAXSNAP) {
    ..   
```

**Review by Matan-B — COMMENTED** [review]
lgtm, make check fails:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/recovery_backend.cc:210:40: error: 'soid' in capture list does not name a variable
    co_await interruptor::async([this, soid, &t] {
                                       ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/recovery_backend.cc:211:44: error: reference to local binding 'soid' declared in enclosing function 'RecoveryBackend::handle_backfill_remove'
      pg.remove_object_maybe_snapmapped(t, soid);
                                           ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/recovery_backend.cc:208:38: note: 'soid' declared here
  for ([[maybe_unused]] const auto& [soid, ver] : m.ls) {
                                     ^
2 errors generated.
```

File: src/crimson/osd/pg.cc (hunk 3 above)
Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc]: All done, I changed the name to `remove_maybe_snapmapped_object()` as the object being removed may not have snaps.

**Review by Matan-B — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins test api

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-08-26_07:29:24-crimson-rados-wip-matanb-crimson-testing-aug-22-distro-crimson-smithi/

**Pull request merged**